### ISSUE TITLE
The bot gets ping timeout after 5-15 minutes.

### ISSUE THREAD

**Issue by afreakk** [Lysxia/roll-your-own-irc-bot#1]
Ive been trying out this bot, and it seems work allright for some time and responds to ping messages, but then suddenly gets stuck on `hGetLine` and then you can see on irc that the bot quits with message `quit (Reason: Ping timeout: 260 seconds)`.

**Comment by Lysxia** [comment]
Thanks for the report!

I think I've encountered similar issues before but never got around to troubleshoot that. It might be caused by the network you're on, in which case retrying might be the only option.

Asking for clarification:

1. Does it happen after 5-15 minutes of inactivity, or can it happen even with activity (has the bot stopped receiving messages in the middle of a conversation)?

2. If I understand correctly, the order things happen is that the bot no longer receives messages for some reason, then the next time the server pings it, the bot cannot respond, and after 260 seconds the server disconnects it. Is that right?

3. What happens to the bot after being disconnected? Is it still stuck, or is it killed by an exception?

**Comment by afreakk** [comment]
1. im not sure, ive tried both with some activity and without, it seems to last longer when there is activity, but my samplesize is small, so not sure.
2. Yes, i think so, it gets stuck on hGetLine, so the two possibilities i guess is the server not receiving my last pong, or im not able to get the servers next ping. (the timing of quit and timeout message from server makes it seem like it did not receive my last PONG)
3. it just gets stuck.

this is me debugging it a bit, ive added timestamp and tried returning \r instead of \r\n because thats what freenode does:
```
...(more ping/pong above here, but cut it short)...
"Getting line."
"Got line."
[15:21:47]"PING :cherryh.freenode.net\r"
[15:21:47]"PONG :cherryh.freenode.net\r"
"Getting line."
"Got line."
[15:24:07]"PING :cherryh.freenode.net\r"
[15:24:07]"PONG :cherryh.freenode.net\r"
"Getting line."
^C⏎    
```
(the bot did not exit, i had to ctr-c to exit myself)
this is some context for the extra lines i print out:
```
    print "Getting line."
    line <- hGetLine h
    print "Got line."
    makeTime >>= putStr 
    print line
```
And this is the log of observing the bot in a channel:
```
 15:14          * | YoutubeTitle joined #shookones
 15:28          * | YoutubeTitle quit (Reason: Ping timeout: 260 seconds)
```

**Comment by Lysxia** [comment]
Yes this seems quite hard to debug. But here are some more ideas for further testing if you have the energy for it:

- Test a similar bot in another language (to know whether the issue is with Haskell).
- Make a little local server to connect your bot to on localhost (thus cutting out the network/ISP).
- Test on a different network, with a different machine.
- It will probably do nothing, but try adding `hFlush h` before `hGetLine h`.

(On my side I've had the bot connected for at least an hour with no issues.)

What's your version of `base` and `network`, and of your OS?

**Comment by afreakk** [comment]
base 4.12.0.0
network 2.8.0.1
I use arch btw :) 
I installed stack through nix though.

**Comment by afreakk** [comment]
On my laptop, with the same versions of base and network, and also using archlinux,
it works, and i dont get ping timeout, even after 6-7 hours.
So the difference there is mostly hardware..

**Comment by afreakk** [comment]
Ok, this code in Rust also ends up with getting `Ping timeout` after ~15 minutes.
https://gist.github.com/afreakk/04988f375bbeb2cb2493e3614883d879
So its not a haskell issue it seems.
Edit: now the rust bot has been running for an hour, so not sure, it seems more stable.
Edit2: no, i was probably just connected to a more lenient server when i didnt disconnect, the rust code can only stay alive for ~15 minutes on tolkien.freenode.net:6667
Edit3: I inspected the packages with wireshark, and every ping **is** responded with a pong, but then the pings stops happening and i get ping timeout.
(when i run irssi on the same computer, the pings continue forever, and I do not get ping timeout)
